libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp

```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
                 last, dest_first);
     }
 }}}    // namespace hpx::parallel::v1
+
+
+// creat new APIs, tag_invoke overloads.
+namespace hpx {
+    ///////////////////////////////////////////////////////////////////////////
+    // CPO for hpx::rotate
+    HPX_INLINE_CONSTEXPR_VARIABLE struct rotate_t final
+      : hpx::functional::tag_fallback<rotate_t>
+    {  
+        // clang-format off
+        template <typename FwdIter, 
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_forward_iterator<FwdIter>::value
+            )>
+        // clang-format on
+        friend FwdIter tag_fallback_dispatch(
+            hpx::rotate_t, FwdIter first, FwdIter new_first, FwdIter last)
+        {
+            static_assert(hpx::traits::is_forward_iterator<FwdIter>::value,
+                "Requires at least forward iterator.");
+
+            return hpx::parallel::v1::detail::rotate<FwdIter>().call2(
+                hpx::execution::sequenced_policy{}, first, new_first, last);
+
+        }
+        // clang-format off
+        template <typename ExPolicy, typename FwdIter,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_forward_iterator<FwdIter>::value
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy, FwdIter>::type
+        tag_fallback_dispatch(
+            hpx::rotate_t, ExPolicy&& policy, FwdIter first, FwdIter new_first, 
+                FwdIter last)
+        {
+            static_assert((hpx::traits::is_forward_iterator<FwdIter>::value),
+                "Requires at least forward iterator.");
+
+            typedef std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
+                    !hpx::traits::is_bidirectional_iterator<FwdIter>::value>
+                is_seq;
+
+            return parallel::util::get_second_element(
+                hpx::parallel::v1::detail::rotate<hpx::parallel::util::in_out_result<FwdIter,FwdIter>>()
+                .call2(std::forward<ExPolicy>(policy), is_seq(), first, new_first, last));
```

Comment:
I'm sure our clang-format will complain about this, please always format the files you changed before committing.